File: typescript/examples/prompt-templates/multimodal.ts
import { llm } from 'mirascope';

const imagePromptTemplate = llm.definePromptTemplate<{
  bookCover: llm.Image | string | Uint8Array;
}>`
  What book is this? {{ bookCover:image }}
`;

const audioPromptTemplate = llm.definePromptTemplate<{
  audio: llm.Audio | string | Uint8Array;
}>`
  Analyze this audio recording: {{ audio:audio }}
`;

const videoPromptTemplate = llm.definePromptTemplate<{
  video: llm.Video | string | Uint8Array;
}>`
  Summarize this video: {{ video:video }}
`;

const documentPromptTemplate = llm.definePromptTemplate<{
  doc: llm.Document | string | Uint8Array;
}>`
  Review this document: {{ doc:document }}
`;

const imagesPromptTemplate = llm.definePromptTemplate<{
  covers: Array<llm.Image | string | Uint8Array>;
}>`
  Compare these book covers: {{ covers:images }}
`;

const audiosPromptTemplate = llm.definePromptTemplate<{
  clips: Array<llm.Audio | string | Uint8Array>;
}>`
  Compare these audio clips: {{ clips:audios }}
`;

const videosPromptTemplate = llm.definePromptTemplate<{
  clips: Array<llm.Video | string | Uint8Array>;
}>`
  Do these video clips remind you of any book? {{ clips:videos }}
`;

const documentsPromptTemplate = llm.definePromptTemplate<{
  docs: Array<llm.Document | string | Uint8Array>;
}>`
  Review these documents: {{ docs:documents }}
`;

const mixedMediaPromptTemplate = llm.definePromptTemplate<{
  cover: llm.Image | string | Uint8Array;
  narration: llm.Audio | string | Uint8Array;
  docs: Array<llm.Document | string | Uint8Array>;
}>`
  Analyze this multimedia presentation:
  - Cover image: {{ cover:image }}
  - Audio narration: {{ narration:audio }}
  - Supporting documents: {{ docs:documents }}
  What is the main theme?
`;

File: python/examples/prompt_templates/multimodal.py
from mirascope import llm


@llm.prompt("What book is this? {{ book_cover:image }}")
def image_prompt_template(book_cover: llm.Image | str | bytes): ...


@llm.prompt("Analyze this audio recording: {{ audio:audio }}")
def audio_prompt_template(audio: llm.Audio | str | bytes): ...


@llm.prompt("Review this document: {{ doc:document }}")
def document_prompt_template(doc: llm.Document | str | bytes): ...


@llm.prompt("Compare these book covers: {{ covers:images }}")
def images_prompt_template(covers: list[llm.Image | str | bytes]): ...


@llm.prompt("Compare these audio clips: {{ clips:audios }}")
def audios_prompt_template(clips: list[llm.Audio | str | bytes]): ...


@llm.prompt("Review these documents: {{ docs:documents }}")
def documents_prompt_template(docs: list[llm.Document | str | bytes]): ...


@llm.prompt(
    """
    Analyze this multimedia presentation:
    - Cover image: {{ cover:image }}
    - Audio narration: {{ narration:audio }}
    - Supporting documents: {{ docs:documents }}
    What is the main theme?
    """
)
def mixed_media_prompt_template(
    cover: llm.Image | str | bytes,
    narration: llm.Audio | str | bytes,
    docs: list[llm.Document | str | bytes],
): ...
